### ISSUE TITLE
prefetch parameter in DataParameter unused

### ISSUE THREAD

**Issue by siebenkopf** [BVLC/caffe#4100]
I have lately run into issues when I tried to implement my own data layer. As the data processing is relatively slow, I wanted to increase the prefetch count. First, I tried to use the ``prefetch`` parameter in
https://github.com/BVLC/caffe/blob/master/src/caffe/proto/caffe.proto#L656
However, this parameter is used nowhere, but rather the PREFETCH_COUNT is a hard-coded constant:
https://github.com/BVLC/caffe/blob/master/include/caffe/layers/base_data_layer.hpp#L71

As my validation set is rather large compared to the batch size (19904 images vs. batch size 64), I would like to increase the PREFETCH_COUNT to 311 (311*64>19904), so that the validation data can be prefetched while the network is training.

In fact, I have implemented such an option for the base_data_layer in my private fork, and I might provide a PR if you like. I have added a ``prefetch`` parameter to all other data layer parameters.
Unfortunately, this breaks the current ``BasePrefetchingDataLayer`` API (as I replaced the constant ``PREFETCH_COUNT`` with a member ``prefetch_count_``), so other people that have created their own data layer and want to merge the master branch might get into trouble...

So, what do you think about that?
